### PR TITLE
225 New node tom tom isochrone map

### LINKED ISSUE #225: TomTom Isochrone Map
The node should use the TomTom API to create an Isochrone map https://developer.tomtom.com/blog/build-different/how-use-isochrones-food-delivery/

Depends on AP-20404 to provide the date and time for the traffic model.

### PR DESCRIPTION
**Pull request by wybert**
This will fix #225 

- [x] check the the parameter dependency

### CODE REVIEW

File: knime_extension/src/nodes/spatialnetwork.py [hunk 1]
```diff
@@ -1423,3 +1423,200 @@ def execute(self, exec_context: knext.ExecutionContext, input1, input2):
...
+        [TomTom API key](https://developer.tomtom.com/knowledgebase/platform/articles/how-to-get-an-tomtom-api-key/)
+        is required to authenticate requests to the 
+        [Calculate Reachable Range service](https://developer.tomtom.com/routing-api/documentation/routing/calculate-reachable-range)
+        which is part of the [Routing API](https://developer.tomtom.com/routing-api/documentation/routing/routing-service)
+        provided by TomTom. 
+        To get an API key, click
+        [here.](https://developer.tomtom.com/knowledgebase/platform/articles/how-to-get-an-tomtom-api-key/)
+        For details about the pricing go to the [TomTom pricing page.](https://developer.tomtom.com/store/maps-api)""",
+        default_value="",
+        # TODO: Validate that the API key is present
```
Comment (koettert) [inline knime_extension/src/nodes/spatialnetwork.py:1550]: See todo

File: knime_extension/src/nodes/spatialnetwork.py [hunk 2]
```diff
@@ -1423,3 +1423,200 @@ def execute(self, exec_context: knext.ExecutionContext, input1, input2):
...
+    )
+
+    _COL_GEOMETRY = "Geometry"
+    _COL_ISOCHRONE = "Isochrone"
+
+    def configure(self, configure_context, input_schema_1):
+        self.c_geo_col = knut.column_exists_or_preset(
+            configure_context, self.c_geo_col, input_schema_1, knut.is_geo
+        )
+        # TODO: Return result schema with the origin id and bounds column
```
Comment (koettert) [inline knime_extension/src/nodes/spatialnetwork.py:1568]: See todo

File: knime_extension/src/nodes/spatialnetwork.py [hunk 3]
```diff
@@ -1423,3 +1423,200 @@ def execute(self, exec_context: knext.ExecutionContext, input1, input2):
...
+        tomtom_base_url = "https://api.tomtom.com/routing/1/calculateReachableRange/"
+        # a sample url:
+        # url = """https://api.tomtom.com/routing/1/calculateReachableRange/52.50931,13.42936/json?energyBudgetInkWh=43&avoid=unpavedRoads&vehicleEngineType=electric&constantSpeedConsumptionInkWhPerHundredkm=50,8.2:130,21.3&key=%s"""%tomkey
+        import requests
+        import json
+        from shapely.geometry import Polygon
+
+        c_gdf = knut.load_geo_data_frame(input1, self.c_geo_col, exec_context)
+
+        # TODO: Execute in a loop with progress and cancellation to support a table with multiple input points
```
Comment (koettert) [inline knime_extension/src/nodes/spatialnetwork.py:1582]: See todo

File: knime_extension/src/nodes/spatialnetwork.py [hunk 4]
```diff
@@ -1423,3 +1423,200 @@ def execute(self, exec_context: knext.ExecutionContext, input1, input2):
...
+        import debugpy
+        import logging
+
+        LOGGER = logging.getLogger(__name__)
+        debugpy.listen(5678)
+        LOGGER.error("Waiting for debugger attach")
+        debugpy.wait_for_client()
+        debugpy.breakpoint()
+
+        # TODO: Handle error status codes and output error code and message
```
Comment (koettert) [inline knime_extension/src/nodes/spatialnetwork.py:1611]: See todo

**Review by koettert — CHANGES_REQUESTED** [review]
Also please improve the testflow to include corner cases e.g. no API key present

File: knime_extension/src/nodes/spatialnetwork.py [hunk 5]
```diff
@@ -1423,3 +1423,131 @@ def execute(self, exec_context: knext.ExecutionContext, input1, input2):
...
+
+
+############################################
+# TomTom Isochrone Map Node
+############################################
+@knext.node(
+    name="""TomTom Isochrone Map""",
+    node_type=knext.NodeType.MANIPULATOR,
+    category=__category,
+    icon_path=__NODE_ICON_PATH + "TomTomIsochrone.png",
```
[the remark this answers is not in the document]
Comment (wybert) [reply knime_extension/src/nodes/spatialnetwork.py:1461]: @UrbanGISer Can you help to resolve this?

File: knime_extension/src/nodes/spatialnetwork.py
Comment (koettert) [inline knime_extension/src/nodes/spatialnetwork.py:1]: Adapt the node to be consistent with the Road Network Isochrone Map node e.g. each budget results in a separate result row.

File: knime_extension/src/nodes/spatialnetwork.py (hunk 3 above)
Comment (koettert) [reply knime_extension/src/nodes/spatialnetwork.py:1583]: The TODO is no longer valid. However we need to compute the global centroid for multiple inputs like in the Road Network Isochrone Map.  In addition we need to support multiple budgets.

File: knime_extension/src/nodes/spatialnetwork.py [hunk 6]
```diff
@@ -1423,3 +1423,201 @@ def execute(self, exec_context: knext.ExecutionContext, input1, input2):
...
+        c_gdf = knut.load_geo_data_frame(input1, self.c_geo_col, exec_context)
+
+        # TODO: Execute in a loop with progress and cancellation to support a table with multiple input points
+        # append boundary column to original input table and let the user specify the name of the appended boundary column
+        # TODO: Compute centroid for any none point input geometry
+        URL = (
+            "%s%s,%s/json?timeBudgetInSec=%s&travelMode=%s&traffic=%s&key=%s&routeType=%s&departAt=%s"
+            % (
+                tomtom_base_url,
+                str(c_gdf[self.c_geo_col].get_coordinates()["y"].values[0]),
```
Comment (koettert) [inline knime_extension/src/nodes/spatialnetwork.py:1590]: Compute global centroid and get the coordinates from it.

**Review by koettert — CHANGES_REQUESTED** [review]

**Comment by wybert** [comment]
> Also please improve the testflow to include corner cases e.g. no API key present

What do you mean by that?

File: knime_extension/src/nodes/spatialnetwork.py (hunk 4 above)
Comment (wybert) [reply knime_extension/src/nodes/spatialnetwork.py]: Can you provide an example?

File: knime_extension/src/nodes/spatialnetwork.py (hunk 1 above)
Comment (wybert) [reply knime_extension/src/nodes/spatialnetwork.py:1567]: Can you provide an example?

File: knime_extension/src/nodes/spatialnetwork.py [hunk 7]
```diff
@@ -1423,3 +1423,284 @@ def execute(self, exec_context: knext.ExecutionContext, input1, input2):
...
+
+    _COL_GEOMETRY = "Geometry"
+    _COL_ISOCHRONE = "Time budget (Mins)"
+
+    def configure(self, configure_context, input_schema_1):
+        self.c_geo_col = knut.column_exists_or_preset(
+            configure_context, self.c_geo_col, input_schema_1, knut.is_geo
+        )
+
+        return knext.Schema(
```
Comment (wybert) [inline knime_extension/src/nodes/spatialnetwork.py:1587]: I need some help for this Schema. It still show the WARN, 
`DataSpec generated by configure does not match spec after execution.`
Can you help me understand what's wrong with my configuration?
@koettert

File: knime_extension/src/nodes/spatialnetwork.py [hunk 8]
```diff
@@ -1423,3 +1423,284 @@ def execute(self, exec_context: knext.ExecutionContext, input1, input2):
...
+            [
+                input_schema_1[self.id_col].ktype,
+                knext.int64(),
+                input_schema_1[self.c_geo_col].ktype,
+            ],
+            [
+                self.id_col,
```
Comment (wybert) [inline knime_extension/src/nodes/spatialnetwork.py:1594]: I need some help for this Schema. It still show the WARN, 
`DataSpec generated by configure does not match spec after execution.`
Can you help me understand what's wrong with my configuration?
@koettert

File: knime_extension/src/nodes/spatialnetwork.py [hunk 9]
```diff
@@ -1423,3 +1423,273 @@ def execute(self, exec_context: knext.ExecutionContext, input1, input2):
...
+
+    timeout = knext.IntParameter(
+        "Request timeout in seconds",
+        "The maximum time in seconds to wait for the request to the TomTom API to succeed.",
+        120,
+        min_value=1,
+        is_advanced=True,
+    )
+
+    _COL_GEOMETRY = "Geometry"
```
Comment (koettert) [inline knime_extension/src/nodes/spatialnetwork.py:1579]: Lower case to be consistent with the other nodes

File: knime_extension/src/util/knime_utils.py [hunk 10]
```diff
@@ -769,3 +769,13 @@ def get_env_path():
 #         ]["weight"]
 
 #     return adjust_list_ref
+
```
Comment (koettert) [inline knime_extension/src/util/knime_utils.py:772]: Can you add a placeholder parameter to the method which contains the default value "your api key here" which should also be checked in addition to empty

File: knime_extension/src/nodes/spatialnetwork.py (hunk 8 above)
Comment (koettert) [reply knime_extension/src/nodes/spatialnetwork.py:1594]: Check that the id column is set e.g.
`
        self.id_col = knut.column_exists_or_preset(
            configure_context, self.id_col, input_schema_1, knut.is_numeric_or_string
        )
`

File: knime_extension/src/nodes/spatialnetwork.py [hunk 11]
```diff
@@ -1423,3 +1423,273 @@ def execute(self, exec_context: knext.ExecutionContext, input1, input2):
...
+    def configure(self, configure_context, input_schema_1):
+        self.c_geo_col = knut.column_exists_or_preset(
+            configure_context, self.c_geo_col, input_schema_1, knut.is_geo
+        )
+
+        return knext.Schema(
+            [
+                input_schema_1[self.id_col].ktype,
+                knext.int64(),
+                input_schema_1[self.c_geo_col].ktype,
```
Comment (koettert) [inline knime_extension/src/nodes/spatialnetwork.py:1591]: Shouldn't you return here the Geo(Polygon) type and not the input column type?

File: knime_extension/src/nodes/spatialnetwork.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (koettert) [reply knime_extension/src/nodes/spatialnetwork.py:1461]: @UrbanGISer icon is still missing :-)

**Review by koettert — CHANGES_REQUESTED** [review]

File: knime_extension/src/nodes/spatialnetwork.py (hunk 11 above)
Comment (wybert) [reply knime_extension/src/nodes/spatialnetwork.py:1591]: It has TYPE_MULTI_POLYGON and TYPE_POLYGON, maybe just put it as TYPE_GEO?

Comment (koettert) [reply knime_extension/src/nodes/spatialnetwork.py]: So we should return TYPE_GEO in configure.

File: knime_extension/src/nodes/spatialnetwork.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (UrbanGISer) [reply knime_extension/src/nodes/spatialnetwork.py:1461]: added

File: knime_extension/src/nodes/spatialnetwork.py (hunk 9 above)
Comment (wybert) [reply knime_extension/src/nodes/spatialnetwork.py:1579]: It seems like the upper cases has been used in this file everywhere

File: knime_extension/src/nodes/spatialnetwork.py [hunk 12]
```diff
@@ -1423,3 +1423,285 @@ def execute(self, exec_context: knext.ExecutionContext, input1, input2):
...
+        include_row_key=False,
+        include_none_column=False,
+    )
+
+    # time_budget_in_min = knext.IntParameter(
+    #     "Time budget in minutes",
+    #     """Specifies the time budget for the isochrone map in minutes. The time budget determines the range of
+    #     the isochrone, representing the maximum travel time from the provided origin point.""",
+    #     default_value=15,
+    # )
```
Comment (koettert) [inline knime_extension/src/nodes/spatialnetwork.py:1515]: Please remove

File: knime_extension/src/nodes/spatialnetwork.py [hunk 13]
```diff
@@ -1423,3 +1423,285 @@ def execute(self, exec_context: knext.ExecutionContext, input1, input2):
...
+
+    timeout = knext.IntParameter(
+        "Request timeout in seconds",
+        "The maximum time in seconds to wait for the request to the TomTom API to succeed.",
+        120,
+        min_value=1,
+        is_advanced=True,
+    )
+
+    _COL_GEOMETRY = "Geometry"
```
Comment (koettert) [inline knime_extension/src/nodes/spatialnetwork.py:1579]: Use new _DEF_GEO_COL_NAME_

File: knime_extension/src/nodes/spatialnetwork.py [hunk 14]
```diff
@@ -1423,3 +1423,285 @@ def execute(self, exec_context: knext.ExecutionContext, input1, input2):
...
+                knut.check_canceled(exec_context)
+                loop_i += 1
+                iso_map_list.append([id_, time_budget, bounds_polygon])
+        gdf = gp.GeoDataFrame(
+            iso_map_list, columns=[self.id_col, self._COL_ISOCHRONE, self._COL_GEOMETRY]
+        )
+        gdf.set_geometry(self._COL_GEOMETRY, inplace=True)
+        gdf.crs = c_gdf.crs
+
+        # URL = (
```
Comment (koettert) [inline knime_extension/src/nodes/spatialnetwork.py:1670]: Remove commented code

**Review by koettert — CHANGES_REQUESTED** [review]

**Review by koettert — CHANGES_REQUESTED** [review]

**Comment by koettert** [comment]
@UrbanGISer can you please provide another icon. Unfortunately we are not allowed to use the original TomTom icon as described here: https://www.tomtom.com/legal/en_gb/guidelines-trademarks-copyrights/

**Pull request merged**